Linting dapp-agoric-basics

mujahidkay opened the pull request:
### Description

This PR enables `tsc` and fixes lint errors for this dapp. Majority of the code is typed now apart from a few instances where I either had to self-define some types or leave them as TO-DOs for future. 

Have segregated the types of files by commits so can be reviewed commit by commit.

turadg commented on contract/test/mintStable.js:
```diff
@@ -45,7 +52,7 @@ export const makeStableFaucet = ({
    * Mint a payment in the fee token.
    *
-   * @param {bigint} value the associated brand is implicitly the fee brand
+   * @param {bigint | any[] | CopySet<any> | CopyBag<any>} value the associated brand is implicitly the fee brand
```
this shouldn't change. you can't meeting a Payment for a natural number with these input types. You'll have to narrow the type at the callsite instead

turadg commented on contract/test/test-postalSvc.js:
```diff
@@ -28,8 +27,50 @@ import {
+/** @typedef {Awaited<ReturnType<makeNodeBundleCache>>} BundleCache */
+
+/**
+ * @typedef {{
```
the work is done now so okay to leave, but we generally define the test context by whatever the return value is of a function that makes the context.

turadg commented on ui/src/components/DisplayAmount.tsx:
```diff
@@ -33,7 +33,7 @@ export const AmountValue = ({
   allowPopup,
 }: {
-  amount: Amount;
+  amount: Amount<'set'> | Amount<'nat'> | Amount<'copyBag'> | Amount<'copySet'>;
```
can it really display all those amounts?

mujahidkay replied:
on a second look `Amount<'set'>['value'] | Amount<'copyBag'>['value']` is all it needs. I'll adjust accordingly.

turadg commented on ui/src/components/PurseAmountInput.tsx:
```diff
@@ -1,12 +1,13 @@
 
 type Props = {
   purse: PurseJSONState<AssetKind>;
+  onChangeInput: (amount: {brand: globalThis.Brand, value: unknown} | null) => void;
```
make this optional so you don't have to give it a no-op function when there's nothing to do

mujahidkay replied:
I see. Good catch.

turadg commented on ui/src/components/mint/MintTickets.tsx:
```diff
@@ -26,7 +26,7 @@ const makeOffer = (
   ];
   const choiceBag = makeCopyBag(choices);
-  const ticketAmount = AmountMath.make(brands.Ticket, choiceBag);
+  const ticketAmount = AmountMath.make((brands.Ticket as Brand), choiceBag);
```
better to type the `brands` object such that all its properties are Brand

turadg reviewed (commented):
Great progress!

mujahidkay commented on contract/tools/ui-kit-goals/queryKit.js:
```diff
@@ -101,8 +108,8 @@ export const makeWalletView = (addr, { query, vstorage }) => {
- * @param {import('./batchQuery.js').VStorage} vstorage
- * @param {import('@endo/marshal').Marshal<string | null>} [m]
+ * @param {VStorage} vstorage
+ * @param {Marshal<any>} [m]
```
`m.fromCapData` expects a `string`. I believe I can change this to:
```
@param {Marshal<string>} [m]
```

mujahidkay commented on contract/scripts/deploy-contract.js:
I'll rebase and adjust all commits accordingly. Thanks!

mujahidkay commented on contract/src/sell-concert-tickets.contract.js:
```diff
@@ -46,11 +49,12 @@ const multiply = (amount, n) => {
  *
  * @param {Amount} sum
- * @param {[string, bigint]} entry
+ * @param {[Key, bigint]} entry
```
`getCopyBagEntries` returns `[Key, bigint][]`. Does it make more sense to type cast over there?

turadg replied:
yeah, I think so.

mujahidkay commented on contract/src/sell-concert-tickets.contract.js:
```diff
@@ -62,15 +66,16 @@ const addMultiples = (sum, entry, inventory) => {
-    (sum, entry) => addMultiples(sum, entry, inventory),
-    AmountMath.makeEmpty(brand),
-  );
+  let finalAmount = /** @type {Amount} */ (AmountMath.makeEmpty(brand));
```
`makeEmpty` return `Amount<"nat">` which is incompatible with `Amount` returned by `addMultiples`. Hence, the cast.

turadg replied:
Hmm, that's a bug in `makeEmpty`. So go ahead and cast but the amounts and brands should be `copyBag` kind.
```suggestion
  let finalAmount = /** @type {Amount<'copyBag'>} */ (AmountMath.makeEmpty(brand));
```

mujahidkay commented on contract/src/sell-concert-tickets.proposal.js:
```diff
@@ -41,9 +45,9 @@ export const makeTerms = (brand, baseUnit) => {
+ *   brand: { Ticket: Brand };
+ *   issuer: { Ticket: Issuer };
+ *   instance: { sellConcertTickets: Instance }
  * }} SellTicketsSpace
```
I need to put this back in. I removed this because TS wasn't able to pick up `PromiseSpaceOf` as its brought in via an ambient import but I later made it work via triple slash directives.

mujahidkay commented:
> Great progress!

@turadg 
I've addressed the comments except for https://github.com/Agoric/dapp-agoric-basics/pull/67#discussion_r1779021051 . It wouldn't allow me to type cast it to a `Amount<'copyBag'>`. I have kept this as is and specified the kind when defining the `Brand` in `Inventory`.
Comments addressed upto [e192627](https://github.com/Agoric/dapp-agoric-basics/pull/67/commits/e192627563a1467b77ee386a9d37748bbfd95b70)

The last two commits are just cleanups. I will rebase and squash commits before merging the PR (as per your suggestion).

turadg commented on contract/src/platform-goals/start-governed-contract.js:
```diff
@@ -55,12 +54,14 @@ export const ParamTypes = /** @type {const} */ ({
  *   issuerKeywordRecord?: IssuerKeywordRecord;
  *   terms: Record<string, unknown>;
- *   privateArgs: StartParams<SF>['privateArgs'];
+ *   privateArgs?: StartParams<SF> extends { privateArgs: unknown }
```
What was the problem here? `StartParams` should figure this out itself.

mujahidkay replied:
`Type '"privateArgs"' cannot be used to index type 'StartParams<SF>'` if its kept as is. Couldn't think of a way out of it - any suggestions?

turadg replied:
not off hand. please put an `XXX` comment to indicate this is some tech debt to clean up later.

https://github.com/Agoric/agoric-sdk/wiki/Coding-Style#code-comments

turadg commented on contract/src/swaparoo.contract.js:
```diff
@@ -146,9 +155,11 @@ export const start = async (zcf, privateArgs, baggage) => {
-      return M.splitRecord({
-        give: M.splitRecord(givePattern),
-      });
+      return harden(
```
doesn't `splitRecord` return a hardened object?

mujahidkay replied:
You are right. It does. Will fix.

turadg commented on contract/test/test-sell-concert-tickets-contract.js:
```diff
@@ -106,10 +116,10 @@ test('Start the contract', async t => {
- * @param {ERef<import('@agoric/zoe/src/zoeService/utils').Instance<AssetContractFn>>} instance
- * @param {Purse} purse
+ * @param {ERef<Instance<AssetContractFn>>} instance
+ * @param {Purse<"nat">} purse
```
nit: use single quotes

turadg commented on contract/test/test-vote-by-committee.js:
```diff
@@ -15,10 +15,22 @@ import { mockWalletFactory, seatLike } from './wallet-tools.js';
+ * @import {ParamChangesOfferArgs} from '@agoric/inter-protocol/src/econCommitteeCharter.js';
+ */
+
+
```
please address all the lint warnings

mujahidkay replied:
yeah, this slipped in unintentionally. :/

turadg commented on contract/tools/ui-kit-goals/marshalTables.js:
```diff
@@ -9,12 +9,6 @@
-/**
- * The null slot indicates that identity is not intended to be preserved.
- *
- * @typedef { string | null } WildSlot
```
oic, `WildSlot` predates this PR.

I think it's helpful to have this type named so we can keep the explanation about what `null` indicates.

To find a better name I went searching in agoric-sdk and it turns out that also has `WildSlot`.
https://github.com/Agoric/agoric-sdk/blob/2d48e143f54911897e8345e1094ae6d19599bef2/multichain-testing/tools/marshalTables.js#L15

@dckc what's the idea behind the name? WDYT of something like `OptionalSlot` or `MaybeSlot` or `BoardSlot`?

mujahidkay replied:
"is not intended to be _preserved_" maybe `TempSlot` as in, its temporary. I'm fine with whatever you agree on and decide.

dckc replied:
I suppose `WildSlot` meant "either a slot or a wildcard". I like `MaybeSlot`. I'm a fan of haskell and the [Typeclassopedia](https://wiki.haskell.org/Typeclassopedia). But when in Rome... I just checked [some typescript docs](https://www.typescriptlang.org/docs/handbook/utility-types.html); there I see "optional". So `OptionalSlot`, please.
